{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "bd78a1c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GGCGTTCAGGCA', 'AAGAATCAGTCA', 'CAAGGAGTTCGC', 'CACGTCAATCAC', 'CAATAATATTCG']\n",
      "3\n",
      "1\n",
      "3\n",
      "2\n",
      "1\n",
      "3\n",
      "2\n",
      "2\n",
      "1\n",
      "2\n",
      "1\n",
      "3\n",
      "1\n",
      "1\n",
      "2\n",
      "2\n",
      "2\n",
      "1\n",
      "1\n",
      "1\n",
      "2\n",
      "1\n",
      "1\n",
      "2\n",
      "2\n",
      "2\n",
      "2\n",
      "1\n",
      "1\n",
      "2\n",
      "1\n",
      "2\n",
      "0\n",
      "1\n",
      "1\n",
      "2\n",
      "0\n",
      "0\n",
      "2\n",
      "1\n",
      "2\n",
      "0\n",
      "2\n",
      "1\n",
      "2\n",
      "0\n",
      "0\n",
      "2\n",
      "3\n",
      "1\n",
      "3\n",
      "0\n",
      "0\n",
      "2\n",
      "2\n",
      "2\n",
      "2\n",
      "0\n",
      "0\n",
      "2\n",
      "CAG\n",
      "CAG\n",
      "CAA\n",
      "CAA\n",
      "CAA\n"
     ]
    }
   ],
   "source": [
    "def CalHammingDistance(seq1, seq2):\n",
    "    assert len(seq1) == len(seq2)\n",
    "    dist = 0\n",
    "    for i in range(len(seq1)):\n",
    "        if seq1[i] != seq2[i]:\n",
    "            dist += 1\n",
    "    return dist\n",
    "\n",
    "def GenerateProfileMat(dna):\n",
    "    seqNum = float(len(dna))\n",
    "    nucleotide = ['A', 'C', 'G', 'T']\n",
    "    matrix = []\n",
    "    for i in range(len(dna[0])):\n",
    "        base_i = [seq[i] for seq in dna]\n",
    "        # print base_i\n",
    "        colProfile = [float(base_i.count(n))/seqNum for n in nucleotide]\n",
    "        matrix.append(colProfile)\n",
    "    return [list(i) for i in zip(*matrix)]\n",
    "\n",
    "def GenerateProfileMatWithPesudoCount(dna):\n",
    "    seqNum = float(len(dna))\n",
    "    nucleotide = ['A', 'C', 'G', 'T']\n",
    "    matrix = []\n",
    "    for i in range(len(dna[0])):\n",
    "        base_i = [seq[i] for seq in dna]\n",
    "        # print base_i\n",
    "        # print \"seqNum = \", seqNum + 1\n",
    "        colProfile = [float(base_i.count(n) + 1)/seqNum + 4 for n in nucleotide]\n",
    "        matrix.append(colProfile)\n",
    "    return [list(i) for i in zip(*matrix)]\n",
    "\n",
    "\n",
    "def GenerateProfileDict(matrix, j):\n",
    "    return {'A': matrix[0][j], 'C': matrix[1][j], 'G': matrix[2][j], 'T': matrix[3][j]}\n",
    "\n",
    "\n",
    "# def ScoreMotifs(motifs):\n",
    "def score(motifs):\n",
    "    '''Returns the score of the dna list motifs.'''\n",
    "    score = 0\n",
    "    for i in range(len(motifs[0])):\n",
    "        motif = ''.join([motifs[j][i] for j in range(len(motifs))])\n",
    "        # print [homogeneous*len(motif) for homogeneous in 'ACGT']\n",
    "        # Calculate the min score between motif and [AAAAA, CCCCC, GGGGG, TTTTT]\n",
    "        # avoiding find the consensus strings\n",
    "        a = min([CalHammingDistance(motif, homogeneous*len(motif)) for homogeneous in 'ACGT'])\n",
    "        print(a)\n",
    "        score += a\n",
    "    return score\n",
    "\n",
    "\n",
    "def FindPortableKmer(text, k, matrix):\n",
    "    # print text\n",
    "    maxPr = 0\n",
    "    prList = []\n",
    "    for i in range(len(text)-k+1):\n",
    "        KmerPr = 1\n",
    "        pattern = text[i:i+k]\n",
    "        # print pattern\n",
    "        for j in range(len(pattern)):\n",
    "            profile = GenerateProfileDict(matrix, j)\n",
    "            # print profile\n",
    "            KmerPr *= profile[pattern[j]]\n",
    "        # print \"KmerPr =\", KmerPr\n",
    "        prList.append(KmerPr)\n",
    "    i = prList.index(max(prList))\n",
    "    maxKmer = text[i:i+k]\n",
    "    return maxKmer\n",
    "\n",
    "\n",
    "\n",
    "def GreedyMotifSearch(dna, k, t, PesudoCount = False):\n",
    "    BestMotifs = [seq[0:k] for seq in dna]\n",
    "    # print BestMotifs\n",
    "    # profileMat = GenerateProfileMat(BestMotifs)\n",
    "    seq = dna[0]\n",
    "    # print seq\n",
    "    for i in range(len(seq) - k + 1):\n",
    "        kmers = seq[i:i+k]\n",
    "        # print \"Kmer = \", kmers\n",
    "        motifs = [kmers]\n",
    "        for j in range(1, len(dna)):\n",
    "            if PesudoCount == True:\n",
    "                mat = GenerateProfileMatWithPesudoCount(motifs)\n",
    "            else:\n",
    "                mat = GenerateProfileMat(motifs)\n",
    "            # print \"Dna =\", dna[j]\n",
    "            # print mat\n",
    "            # print \"Find Portable Kmer:\"\n",
    "            tempMotif = FindPortableKmer(dna[j], k, mat)\n",
    "            motifs.append(tempMotif)\n",
    "        if score(motifs) < score(BestMotifs):\n",
    "            BestMotifs = motifs\n",
    "#         print(str(score(motifs))  + '|' + str(score(BestMotifs)))\n",
    "    return BestMotifs\n",
    "\n",
    "\n",
    "with open(\"/home/cruelper/university/bioinfo/src/input.txt\", \"r\") as f:\n",
    "    k = f.readline().strip()\n",
    "    dnaSet = [line.strip() for line in f]\n",
    "    print(dnaSet)\n",
    "res = GreedyMotifSearch(dnaSet, 3, 5, False)\n",
    "for i in res:\n",
    "    print(i) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7843deaf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2d3a176",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
